File: leetcode/79.word-search/main.go
package main

func search(board [][]byte, word string, x int, y int, u int, points [][]bool) bool {
	if x < 0 || x >= len(board) {
		return false
	}
	if y < 0 || y >= len(board[x]) {
		return false
	}
	if board[x][y] != word[u] {
		return false
	}
	if points[x][y] {
		return false
	}
	if u == len(word)-1 {
		return true
	}

	points[x][y] = true
	if search(board, word, x-1, y, u+1, points) {
		return true
	}
	if search(board, word, x+1, y, u+1, points) {
		return true
	}
	if search(board, word, x, y-1, u+1, points) {
		return true
	}
	if search(board, word, x, y+1, u+1, points) {
		return true
	}
	points[x][y] = false

	return false
}

func exist(board [][]byte, word string) bool {
	if len(word) == 0 {
		return false
	}

	var points = make([][]bool, len(board))
	for m := 0; m < len(board); m++ {
		points[m] = make([]bool, len(board[m]))
	}

	for m := 0; m < len(board); m++ {
		for n := 0; n < len(board[m]); n++ {
			if board[m][n] == word[0] {
				if search(board, word, m, n, 0, points) {
					return true
				}
			}
		}
	}

	return false
}

func main() {
	board := [][]byte{
		{'A', 'B', 'C', 'E'},
		{'S', 'F', 'E', 'S'},
		{'A', 'D', 'E', 'E'},
	}
	word := "ABCESEEEFS"
	println(exist(board, word))
}

File: leetcode/79.word-search/main.rs
pub struct Solution;

impl Solution {
    pub fn exist(board: Vec<Vec<char>>, word: String) -> bool {
        let mut points = Vec::new();
        points.resize(board.len(), vec![]);
        for i in 0..board.len() {
            points[i].resize(board[i].len(), false);
        }
        let word = word.chars().collect::<Vec<char>>();

        for i in 0..board.len() {
            for j in 0..board[i].len() {
                if Self::search(&board, &word, i, j, 0, &mut points) {
                    return true;
                }
            }
        }

        false
    }

    fn search(
        board: &Vec<Vec<char>>,
        word: &Vec<char>,
        i: usize,
        j: usize,
        k: usize,
        points: &mut Vec<Vec<bool>>,
    ) -> bool {
        if points[i][j] {
            return false;
        }
        if board[i][j] != word[k] {
            return false;
        }
        if k + 1 == word.len() {
            return true;
        }

        points[i][j] = true;
        if i >= 1 && Self::search(board, word, i - 1, j, k + 1, points) {
            return true;
        }
        if i + 1 < board.len() && Self::search(board, word, i + 1, j, k + 1, points) {
            return true;
        }

        if j >= 1 && Self::search(board, word, i, j - 1, k + 1, points) {
            return true;
        }
        if j + 1 < board[i].len() && Self::search(board, word, i, j + 1, k + 1, points) {
            return true;
        }
        points[i][j] = false;
        return false;
    }
}

fn main() {
    let arr = vec![
        vec!['A', 'B', 'C', 'E'],
        vec!['S', 'F', 'C', 'S'],
        vec!['A', 'D', 'E', 'E'],
    ];
    let b = Solution::exist(arr, "ABCCED".to_owned());
    println!("{:?}", b);
}
